File: samples/replicatedentity-examples/src/main/java/com/example/replicated/countermap/domain/SomeCounterMap.java
package com.example.replicated.countermap.domain;

import com.akkaserverless.javasdk.replicatedentity.ReplicatedCounterMap;
import com.akkaserverless.javasdk.replicatedentity.ReplicatedEntityContext;
import com.example.replicated.countermap.SomeCounterMapApi;
import com.google.protobuf.Empty;

import java.util.AbstractMap.SimpleEntry;
import java.util.Map;
import java.util.stream.Collectors;

public class SomeCounterMap extends AbstractSomeCounterMap {
  @SuppressWarnings("unused")
  private final String entityId;

  public SomeCounterMap(ReplicatedEntityContext context) {
    this.entityId = context.entityId();
  }

  // tag::update[]
  @Override
  public Effect<Empty> increase(
      ReplicatedCounterMap<SomeCounterMapDomain.SomeKey> counterMap,
      SomeCounterMapApi.IncreaseValue command) {
    SomeCounterMapDomain.SomeKey key = // <1>
        SomeCounterMapDomain.SomeKey.newBuilder().setKey(command.getKey()).build();
    return effects()
        .update(counterMap.increment(key, command.getValue())) // <2>
        .thenReply(Empty.getDefaultInstance());
  }

  @Override
  public Effect<Empty> decrease(
      ReplicatedCounterMap<SomeCounterMapDomain.SomeKey> counterMap,
      SomeCounterMapApi.DecreaseValue command) {
    SomeCounterMapDomain.SomeKey key = // <1>
        SomeCounterMapDomain.SomeKey.newBuilder().setKey(command.getKey()).build();
    return effects()
        .update(counterMap.decrement(key, command.getValue())) // <2>
        .thenReply(Empty.getDefaultInstance());
  }

  @Override
  public Effect<Empty> remove(
      ReplicatedCounterMap<SomeCounterMapDomain.SomeKey> counterMap,
      SomeCounterMapApi.RemoveValue command) {
    SomeCounterMapDomain.SomeKey key = // <1>
        SomeCounterMapDomain.SomeKey.newBuilder().setKey(command.getKey()).build();
    return effects()
        .update(counterMap.remove(key)) // <2>
        .thenReply(Empty.getDefaultInstance());
  }
  // end::update[]

  // tag::get[]
  @Override
  public Effect<SomeCounterMapApi.CurrentValue> get(
      ReplicatedCounterMap<SomeCounterMapDomain.SomeKey> counterMap,
      SomeCounterMapApi.GetValue command) {
    SomeCounterMapDomain.SomeKey key = // <1>
        SomeCounterMapDomain.SomeKey.newBuilder().setKey(command.getKey()).build();
    long value = counterMap.get(key); // <2>
    SomeCounterMapApi.CurrentValue currentValue =
        SomeCounterMapApi.CurrentValue.newBuilder().setValue(value).build();
    return effects().reply(currentValue);
  }

  @Override
  public Effect<SomeCounterMapApi.CurrentValues> getAll(
      ReplicatedCounterMap<SomeCounterMapDomain.SomeKey> counterMap,
      SomeCounterMapApi.GetAllValues command) {
    Map<String, Long> values =
        counterMap.keySet().stream() // <3>
            .map(key -> new SimpleEntry<>(key.getKey(), counterMap.get(key)))
            .collect(Collectors.toMap(SimpleEntry::getKey, SimpleEntry::getValue));
    SomeCounterMapApi.CurrentValues currentValues =
        SomeCounterMapApi.CurrentValues.newBuilder().putAllValues(values).build();
    return effects().reply(currentValues);
  }
  // end::get[]
}

File: samples/scala-replicated-examples/src/main/scala/com/example/replicated/countermap/domain/SomeCounterMap.scala
package com.example.replicated.countermap.domain

import com.akkaserverless.scalasdk.replicatedentity.ReplicatedCounterMap
import com.akkaserverless.scalasdk.replicatedentity.ReplicatedEntity
import com.akkaserverless.scalasdk.replicatedentity.ReplicatedEntityContext
import com.example.replicated.countermap
import com.google.protobuf.empty.Empty

// This class was initially generated based on the .proto definition by Akka Serverless tooling.
//
// As long as this file exists it will not be overwritten: you can maintain it yourself,
// or delete it so it is regenerated as needed.

/** A replicated entity. */
class SomeCounterMap(context: ReplicatedEntityContext) extends AbstractSomeCounterMap {


  /** Command handler for "Increase". */
  def increase(currentData: ReplicatedCounterMap[String], increaseValue: countermap.IncreaseValue): ReplicatedEntity.Effect[Empty] =
    effects.error("The command handler for `Increase` is not implemented, yet")

  /** Command handler for "Decrease". */
  def decrease(currentData: ReplicatedCounterMap[String], decreaseValue: countermap.DecreaseValue): ReplicatedEntity.Effect[Empty] =
    effects.error("The command handler for `Decrease` is not implemented, yet")

  /** Command handler for "Remove". */
  def remove(currentData: ReplicatedCounterMap[String], removeValue: countermap.RemoveValue): ReplicatedEntity.Effect[Empty] =
    effects.error("The command handler for `Remove` is not implemented, yet")

  /** Command handler for "Get". */
  def get(currentData: ReplicatedCounterMap[String], getValue: countermap.GetValue): ReplicatedEntity.Effect[countermap.CurrentValue] =
    effects.error("The command handler for `Get` is not implemented, yet")

  /** Command handler for "GetAll". */
  def getAll(currentData: ReplicatedCounterMap[String], getAllValues: countermap.GetAllValues): ReplicatedEntity.Effect[countermap.CurrentValues] =
    effects.error("The command handler for `GetAll` is not implemented, yet")

}
